Kernel/Devices/USB/UHCIController.cpp

```diff
@@ -312,62 +315,191 @@ void UHCIController::start()
+    if (td == nullptr) {
+        return nullptr;
+    }
+
+    u16 max_len = (data_len > 0) ? (data_len - 1) : 0x7ff;
+
+    // Set the token dword
+    td->set_token((max_len << TD_TOKEN_MAXLEN_SHIFT) | ((pipe.data_toggle() ? 1 : 0) << TD_TOKEN_DATA_TOGGLE_SHIFT) | (pipe.endpoint_address() << TD_TOKEN_ENDPOINT_SHIFT) | (pipe.device_address() << TD_TOKEN_DEVICE_ADDR_SHIFT) | (static_cast<u8>(direction)));
+
+    pipe.set_toggle(!pipe.data_toggle());
+
+    // Set some bits in the control/status register
+    if (pipe.type() == USBPipe::Type::Isochronous) {
+        td->set_isochronous();
+    } else {
+        if (direction == PacketID::IN) {
+            td->set_short_packet_detect();
+        }
+    }
+
+    // Set lowsped bit if the device connected to port is a Lowspeed device (proably unlikely...)
+    if (pipe.device_speed() == USBPipe::DeviceSpeed::LowSpeed) {
+        td->set_lowspeed();
+    }
+
+    td->set_active();
+    td->set_error_retry_counter(RETRY_COUNTER_RELOAD);
+
+    return td;
+}
+
+bool UHCIController::create_chain(USBPipe& pipe, PacketID direction, uint32_t buff_addr, size_t max_size, size_t transfer_size, TransferDescriptor** td_chain, TransferDescriptor** last_td)
+{
+    // We need to create `n` transfer descriptors based on the max
+    // size of each transfer (which we've learned from the device already by reading
+    // it's device descriptor, or 8 bytes). Each TD then has its' buffer pointer
+    // set to the inital buffer address + (max_size * index), where index is
+    // the ID of the TD in the chain.
+    size_t byte_count = 0;
+    TransferDescriptor* current_td = nullptr;
+    TransferDescriptor* prev_td = nullptr;
+    TransferDescriptor* first_td = nullptr;
+
+    // Keep creating transfer descriptors while we still have some data
+    while (byte_count < transfer_size) {
+        size_t packet_size = transfer_size - byte_count;
+        if (packet_size > max_size) {
+            packet_size = max_size;
+        }
+
+        current_td = create_transfer_descriptor(pipe, direction, packet_size);
+        if (current_td == nullptr)
+            return false;
+
+        current_td->set_buffer_address(buff_addr + byte_count);
+        byte_count += packet_size;
+
+        if (prev_td != nullptr)
+            prev_td->insert_next_transfer_descriptor(current_td);
+        else
+            first_td = current_td;
+
+        prev_td = current_td;
+    }
+
+    *last_td = current_td;
+    *td_chain = first_td;
+    return true;
+}
+
+void UHCIController::free_descriptor_chain(TransferDescriptor* first_descriptor)
+{
+    TransferDescriptor* descriptor = first_descriptor;
+
+    while (descriptor) {
+        descriptor->free();
+        descriptor = descriptor->next_td();
+    }
+}
+
+size_t UHCIController::submit_control_transfer(USBTransfer& transfer)
+{
+    USBPipe& pipe = const_cast<USBPipe&>(transfer.pipe()); // Short circuit the pipe related to this transfer
```

Comment:
Can we make `USBTransfer` have a non-const version of `pipe()` so we don't have to `const_cast` here?

Reply:
Sure!